### PR TITLE
Adanfo feature registry

### PR DESCRIPTION
**Pull request by adanforth**
### Motivation
Currently, main.cpp is using a vector to keep track of its active features. This is a very cumbersome solution. We introduce a new class, FeatureRegistry, which will behave exactly the same, but will abstract away some of the logic into its methods.

#### Change summary
- Added FeatureRegistry.h/.cpp
- Added TestFeatureRegistry.cpp
- Modified Main.cpp to make use of this new class
- Modified SharedCrtResourceManager to make use of this class as well

### Testing
Small Unit tests have been created to test basic functionality

### NOTE
ConfigShadow Feature currently has an error when enabled, the DC will not shut down gracefully.

### CODE REVIEW

File: source/SharedCrtResourceManager.h [hunk 1]
```diff
@@ -42,7 +43,8 @@ namespace Aws
                 std::shared_ptr<Crt::Mqtt::MqttConnection> connection;
                 aws_allocator *allocator{nullptr};
                 aws_mem_trace_level memTraceLevel{AWS_MEMTRACE_NONE};
-                std::vector<Feature *> *features;
+                Util::FeatureRegistry *features;
```
Comment (marcoemorais-aws) [inline source/SharedCrtResourceManager.h:46]: We want to avoid naked pointers except for C runtime objects such as `aws_allocator` and so forth. Make this a shared pointer to `FeatureRegistry`.

```
std::shared_ptr<Util::FeatureRegistry> features;
```

File: source/SharedCrtResourceManager.h [hunk 2]
```diff
@@ -73,7 +75,7 @@ namespace Aws
                 static const int RETRY = 1;
                 static const int ABORT = 2;
 
-                bool initialize(const PlainConfig &config, std::vector<Feature *> *features);
+                bool initialize(const PlainConfig &config, Util::FeatureRegistry *featureRegistry);
```
Comment (marcoemorais-aws) [inline source/SharedCrtResourceManager.h:78]: Take a shared pointer `std::shared_ptr<Util::FeatureRegistry> features`.

File: source/main.cpp [hunk 3]
```diff
@@ -94,10 +95,11 @@ using namespace Aws::Iot::DeviceClient::SensorPublish;
 
 const char *TAG = "Main.cpp";
 
-vector<Feature *> features;
+// vector<Feature *> features;
```
Comment (marcoemorais-aws) [inline source/main.cpp:98]: Don't leave commented code like this since it will be confusing and more difficult to read. I won't repeat that but it applies in the code that follows. Since we are using version control, we can always examine the previous code without needing to leave it commented in the source file.

File: source/main.cpp [hunk 4]
```diff
@@ -94,10 +95,11 @@ using namespace Aws::Iot::DeviceClient::SensorPublish;
...
+FeatureRegistry features;
```
Comment (marcoemorais-aws) [inline source/main.cpp:99]: Make it a `shared_ptr<FeatureRegistry>` and similar to `SharedCrtResourceManager` initialize in main.cpp.

File: source/main.cpp [hunk 5]
```diff
@@ -456,7 +463,8 @@ int main(int argc, char *argv[])
         LOG_INFO(TAG, "Jobs is enabled");
         jobs = unique_ptr<JobsFeature>(new JobsFeature());
         jobs->init(resourceManager->getConnection(), listener, config.config);
-        features.push_back(jobs.get());
+        features.add(shared_ptr<Feature>(jobs.get()));
```
Comment (marcoemorais-aws) [inline source/main.cpp:466]: Bug. unique_ptr is about transfer of ownership. When you call `.get()` on a unique pointer you get a raw pointer to the pointed to object, but the ownership of the object it points to remains with the original unique_ptr. If you were to pass the unique_ptr to FeatureRegistry then the ownership transfers and you can't refer to that object. 

Instead of using `.get()`, what you want to do here is change all of these `unique_ptr` Feature instances to `shared_ptr` and then pass the shared_ptr to `FeatureRegistry::add`. Both main.cpp and FeatureRegistry will share the reference to the same memory. That memory will be deallocated when the reference count for each goes to 0.

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]

File: source/main.cpp [hunk 6]
```diff
@@ -568,6 +580,9 @@ int main(int argc, char *argv[])
                 break;
             case SIGHUP:
                 resourceManager->dumpMemTrace();
+                FeatureRegistry reg;
```
Comment (marcoemorais-aws) [inline source/main.cpp:583]: This is weird. You have to call stop on the registry with features in it, not this temporary.

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]

File: source/main.cpp (hunk 6 above)
Comment (adanforth) [reply source/main.cpp:583]: That was left overs from testing, I forgot to delete in the push.

File: source/util/FeatureRegistry.cpp [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::shared_ptr<Feature> feature)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (std::find(features.begin(), features.end(), feature) != features.end())
```
Comment (marcoemorais-aws) [inline source/util/FeatureRegistry.cpp:32]: This loop will search for features using `operator==`. In this case that means feature is considered equal when address are equal. I am not sure whether that is what you intended and the behavior of `add` is not symmetric with `get` since the latter matches by name. If you instead wanted to search for feature matching by name, then you need something like this.

```
auto result = std::find_if(
    std::begin(features), 
    std::end(features), 
    [&feature](auto f) { return f->getName() == feature->getName(); }
);
```

File: source/util/FeatureRegistry.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        if (features.at(i)->getName() == name)
+        {
+            features.erase(features.begin() + i);
+            break;
+        }
+    }
+}
+
+int FeatureRegistry::getSize()
```
Comment (marcoemorais-aws) [inline source/util/FeatureRegistry.cpp:55]: 1. Return type of `std::vector::size` is `std::size_t` (or `std::vector::size_type` which is same thing). Similar to the previous comment, use `std::size_t` instead of `int`.
2. Since this class maintains a lock to protect concurrent modification of `features`, then this function also needs to acquire a lock before invoking member function on `features`.

File: source/util/FeatureRegistry.cpp [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+{
+    return features.size();
+}
+
+void FeatureRegistry::stopAll()
+{
+    while (!features.empty())
```
Comment (marcoemorais-aws) [inline source/util/FeatureRegistry.cpp:62]: This logic is strange. Assuming there are no side-effects it will just call `Feature::stop` on the same index in an endless loop.  Based on the name I would expect that this function is symmetric with `startAll` and iterates through all the features and calls `Feature::stop`. If the intent is to both `stop` a feature and do something else (I think that it is), that seems flawed and let's talk about that offline.

Similarly strange is the fact that `startAll` acquires a lock but `stopAll` does not.

File: source/util/FeatureRegistry.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                 * @brief A class to keep track of the features currently running on the device client
+                 */
+                class FeatureRegistry
+                {
+                  public:
+                    FeatureRegistry() = default;
+                    std::shared_ptr<Feature> get(const std::string &name) const;
+                    void add(std::shared_ptr<Feature> feature);
+                    void remove(const std::string &name);
+                    int getSize();
```
Comment (marcoemorais-aws) [inline source/util/FeatureRegistry.h:34]: `getSize()`, `stopAll`, and `startAll` can all be `const` member functions eg add `const` suffix.

File: test/util/TestFeatureRegistry.cpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+using namespace std;
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::DeviceDefender;
+
+class TestFeatureRegistry : public ::testing::Test
+{
+  public:
+    shared_ptr<FeatureRegistry> features;
+    shared_ptr<DeviceDefenderFeature> deviceDefender;
```
Comment (marcoemorais-aws) [inline test/util/TestFeatureRegistry.cpp:20]: You can invent dummy features for this test. Use something like to below to create `FakeFeature` instances with different names so that you can test the api with more than 1 entry in the registry. Overall you need more tests.

```
class FakeFeature : public Feature
{
  public:
    FakeFeature(const std::string& name) :name(name) {}
    void start() override { }
    void stop() override { }
    std::string getName() { return name; }
  private:
     std::string name;
};
```

File: source/main.cpp [hunk 12]
```diff
@@ -447,16 +420,16 @@ int main(int argc, char *argv[])
...
+    // featuresReadWriteLock.lock(); // LOCK
 
 #if !defined(EXCLUDE_JOBS)
-    unique_ptr<JobsFeature> jobs;
+    shared_ptr<JobsFeature> jobs;
     if (config.config.jobs.enabled)
     {
         LOG_INFO(TAG, "Jobs is enabled");
-        jobs = unique_ptr<JobsFeature>(new JobsFeature());
+        jobs = shared_ptr<JobsFeature>(new JobsFeature());
```
Comment (marcoemorais-aws) [inline source/main.cpp:430]: Be consistent and use `make_shared` throughout this file.

File: source/SharedCrtResourceManager.h [hunk 13]
```diff
@@ -42,7 +43,8 @@ namespace Aws
                 std::shared_ptr<Crt::Mqtt::MqttConnection> connection;
                 aws_allocator *allocator{nullptr};
                 aws_mem_trace_level memTraceLevel{AWS_MEMTRACE_NONE};
-                std::vector<Feature *> *features;
+                std::shared_ptr<Util::FeatureRegistry> features;
+                // std::vector<Feature *> *features;
```
Comment (marcoemorais-aws) [inline source/SharedCrtResourceManager.h:47]: Remove commented code, we have version control. I see this in other places in this review, but won't repeat comment.

File: source/util/FeatureRegistry.h [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+                /**
+                 * @brief A class to keep track of the features currently running on the device client
+                 */
+                class FeatureRegistry
+                {
+                  public:
+                    FeatureRegistry() = default;
+                    std::shared_ptr<Feature> get(const std::string &name) const;
+                    void add(std::shared_ptr<Feature> feature);
```
Comment (marcoemorais-aws) [inline source/util/FeatureRegistry.h:32]: Add comments to describe each function.

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]

File: source/SharedCrtResourceManager.h [hunk 15]
```diff
@@ -42,7 +43,7 @@ namespace Aws
                 std::shared_ptr<Crt::Mqtt::MqttConnection> connection;
                 aws_allocator *allocator{nullptr};
                 aws_mem_trace_level memTraceLevel{AWS_MEMTRACE_NONE};
-                std::vector<Feature *> *features;
+                std::shared_ptr<Util::FeatureRegistry> features;
```
Comment (shangabl) [inline source/SharedCrtResourceManager.h:46]: nit: use Util namespace so you can just have <FeatureRegistry>

File: source/main.cpp [hunk 16]
```diff
@@ -94,10 +95,9 @@ using namespace Aws::Iot::DeviceClient::SensorPublish;
 
 const char *TAG = "Main.cpp";
 
-vector<Feature *> features;
+shared_ptr<FeatureRegistry> features = make_shared<FeatureRegistry>();
```
Comment (shangabl) [inline source/main.cpp:98]: Just declare here and then init in the main method.

File: source/main.cpp [hunk 17]
```diff
@@ -131,32 +131,24 @@ bool init(int argc, char *argv[])
...
 
-        for (auto &feature : featuresCopy)
-        {
-            LOGM_DEBUG(TAG, "Attempting shutdown of %s", feature->getName().c_str());
-            feature->stop();
-        }
-
-        resourceManager->dumpMemTrace();
+        LOG_DEBUG(TAG, "Calling stop all");
+        features->stopAll();
```
Comment (shangabl) [inline source/main.cpp:140]: Why not just remove each feature after you stop it? Seems like these are only called together.

File: source/util/FeatureRegistry.cpp [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+#include "FeatureRegistry.h"
+#include "../logging/LoggerFactory.h"
+
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
```
Comment (shangabl) [inline source/util/FeatureRegistry.cpp:13]: Is this used? I'm not seeing it used outside of your unit tests.

File: source/main.cpp (hunk 17 above)
Comment (adanforth) [reply source/main.cpp:140]: There were some logic issues doing it that way, which is how it was originally done. the call to remove requires the lock as well, recursive locks could be used in that case, but I was advised against them. I had issues with the existing logic inside of included handle_feature_stopped, which would call shutdown once the vector was empty which means we couldn't remove the last feature since now we track the features in a vector in FeatureRegistry. However, now that I removed that logic, we could just move the logic of this removeAll() call inside of stopAll(), but I don't see a need to currently. 
To summarize, I made it this way to work around other errors in main.cpp, but those errors are gone and now this remains, but I think keeping it this way is fine.

File: source/util/FeatureRegistry.cpp (hunk 18 above)
Comment (adanforth) [reply source/util/FeatureRegistry.cpp:13]: I will be using get in dynamic reload

File: source/main.cpp (hunk 17 above)
Comment (shangabl) [reply source/main.cpp:140]: Adding an entire new method to a class because of something that no longer exists is generally not recommended.

File: source/util/FeatureRegistry.cpp (hunk 18 above)
Comment (shangabl) [reply source/util/FeatureRegistry.cpp:13]: I would probably just add it in your dynamic reload PR then. Adding a method you will need later that doesn't actually get used does not seem like a good idea. What if AppSec doesn't approve your project or it is never merged for some other reason?

Comment (marcoemorais-aws) [reply source/util/FeatureRegistry.cpp:13]: This method is prolly needed for testing.

File: source/FeatureRegistry.cpp [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+using namespace Aws::Iot::DeviceClient::Util;
+using namespace Aws::Iot::DeviceClient::Logging;
+
+constexpr char FeatureRegistry::TAG[];
+
+std::shared_ptr<Feature> FeatureRegistry::get(const std::string &name) const
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+
+    for (auto &feature : features)
```
Comment (shangabl) [inline source/FeatureRegistry.cpp]: Addressed in Slack

File: source/FeatureRegistry.cpp [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::string name, std::shared_ptr<Feature> featureToAdd)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+
+    for (auto &feature : features)
```
Comment (shangabl) [inline source/FeatureRegistry.cpp]: same

File: source/FeatureRegistry.cpp [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        }
+    }
+
+    features[name] = featureToAdd;
+}
+
+void FeatureRegistry::remove(const std::string &name)
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    for (auto &feature : features)
```
Comment (shangabl) [inline source/FeatureRegistry.cpp]: same

File: source/FeatureRegistry.cpp [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+    return features.size();
+}
+
+void FeatureRegistry::stopAll()
+{
+    std::lock_guard<std::mutex> lock(featuresLock);
+
+    for (auto feature : features)
```
Comment (shangabl) [inline source/FeatureRegistry.cpp:57]: You shouldn't need two for-loops here.

File: test/util/TestFeatureRegistry.cpp [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+
+    void SetUp() override
+    {
+        features = make_shared<FeatureRegistry>();
+        feature1 = make_shared<FakeFeature>("feature-1");
+        feature2 = make_shared<FakeFeature>("feature-2");
+        feature3 = make_shared<FakeFeature>("feature-3");
+    }
+
+    void TearDown() override { features = make_shared<FeatureRegistry>(); }
```
Comment (shangabl) [inline test/util/TestFeatureRegistry.cpp:56]: I don't think this is needed? Could be wrong though.

the features shared_ptr will be reinitialized before each test by the SetUp method.

File: test/util/TestFeatureRegistry.cpp [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+};
+
+TEST_F(TestFeatureRegistry, AddFeaturesToRegistry)
```
Comment (shangabl) [inline test/util/TestFeatureRegistry.cpp:59]: Add comments to explain each test.

Also, I think it would be better to have each test only doing one thing. This is testing to make sure add is idempotent and that multiple distinct features can be added and that the size will be correct... I think you'd actually want to test that the size was correct and that it contained all the correct features. But perhaps split the idempotency test off?

File: test/util/TestFeatureRegistry.cpp [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+    ASSERT_EQ(1, features->getSize());
+
+    features->add(feature2->getName(), feature2);
+    ASSERT_EQ(2, features->getSize());
+
+    features->add(feature3->getName(), feature3);
+    ASSERT_EQ(3, features->getSize());
+}
+
+TEST_F(TestFeatureRegistry, GetNonExistentFeature)
```
Comment (shangabl) [inline test/util/TestFeatureRegistry.cpp:74]: Not sure I like this test as there is nothing is added in the first place. Any string value would return nullptr.

File: test/util/TestFeatureRegistry.cpp [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+
+TEST_F(TestFeatureRegistry, StartAllFeatures)
+{
+    features->add(feature1->getName(), feature1);
+    features->add(feature2->getName(), feature2);
+    features->add(feature3->getName(), feature3);
+
+    features->startAll();
+
+    ASSERT_TRUE(feature1->isStarted());
```
Comment (shangabl) [inline test/util/TestFeatureRegistry.cpp:106]: I think there is an ASSERT_ALL where you could group these together to be more readable.

File: source/FeatureRegistry.cpp [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+        feature.second->stop();
+    }
+
+    for (auto &feature : features)
+    {
+        feature.second = nullptr;
+    }
+}
+
+void FeatureRegistry::startAll() const
```
Comment (RogerZhongAWS) [inline source/FeatureRegistry.cpp:69]: you should check if the value is nullptr for startAll and stopAll

File: source/FeatureRegistry.cpp [hunk 28]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        features[name] = featureToAdd;
+    }
+    else
+    {
+        LOGM_WARN(TAG, "Attemped to add %s to Feature Registry despite it already existing", name.c_str());
+    }
+}
+
+void FeatureRegistry::remove(const std::string &name)
```
Comment (RogerZhongAWS) [inline source/FeatureRegistry.cpp:38]: Since we now intend for nullptr to mean the feature is disabled, the meaning of the function name "remove" is no longer accurate. Consider renaming to "disable" or even "setNull"

File: source/FeatureRegistry.cpp [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+    std::lock_guard<std::mutex> lock(featuresLock);
+    if (features.count(name))
+    {
+        return features.at(name);
+    }
+    LOGM_WARN(TAG, "Feature, %s, not found in current registry", name.c_str());
+    return nullptr;
+}
+
+void FeatureRegistry::add(std::string name, std::shared_ptr<Feature> featureToAdd)
```
Comment (RogerZhongAWS) [inline source/FeatureRegistry.cpp:25]: It looks like we would need a "set" function to handle the cases where you enable a previously disabled feature. add() is only able to function upon startup since the features map hasn't been populated yet. We should discuss with Marco about whether we should create a new function or modify this one.

**Review by RogerZhongAWS — CHANGES_REQUESTED** [review]

File: test/util/TestFeatureRegistry.cpp (hunk 26 above)
Comment (adanforth) [reply test/util/TestFeatureRegistry.cpp:106]: Not sure if I see it on here: https://google.github.io/googletest/reference/assertions.html

File: source/FeatureRegistry.cpp (hunk 28 above)
Comment (adanforth) [reply source/FeatureRegistry.cpp:38]: I think that from a users standpoint, it is essentially removed? They will be able to try and get the feature and we return that it is not there. There are no ways that we add more or less things into the registry after main.cpp runs. We simple have placeholder keys (this will be useful for dynamic reload) but for all intents and purposes, it is removed.

File: source/FeatureRegistry.cpp (hunk 29 above)
Comment (adanforth) [reply source/FeatureRegistry.cpp:25]: I think add serves a good purpose of literally adding features to the registry. Maybe a new method set() or something similar will be useful for dynamic reload. However, I will add this in the dynamic reload PR because it serves no purpose as of now.

File: source/FeatureRegistry.cpp (hunk 28 above)
Comment (RogerZhongAWS) [reply source/FeatureRegistry.cpp:38]: Yes, but the intended audience for the source code is developers, not users. If you have a class that is a wrapper around a container, its function names and behaviors are expected to be correspond with other examples of wrappers around containers.

**Review by shangabl — DISMISSED** [review]

**Review by RogerZhongAWS — DISMISSED** [review]

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]

**Review by marcoemorais-aws — DISMISSED** [review]

**Review by RogerZhongAWS — APPROVED** [review]

**Review by marcoemorais-aws — CHANGES_REQUESTED** [review]
@adanforth 

## Issue
The normal shutdown path for device client will go through `shutdown()` function in `source/main.cpp`. In testing this change I discovered that the `shutdown()` function would hang waiting for resource cleanup as part of `aws_http_library_clean_up()`.

## Background
After digging a little deeper I learned the following things.

- The `aws_http_library_clean_up()` function is called from both the `~SecureTunnelingFeature()` destructor and the `~ApiHandle()` destructor in the c++ sdk. The ApiHandle is a member of SharedCrtResourceManager and used for communication with cloud via MQTT. By design, the lifetime of the ApiHandle and SharedCrtResourceManager are tied to the lifetime of the device client. As a result, we can remove the invocation of `aws_http_library_clean_up()`  from `~SecureTunnelingFeature()` destructor, since the cleanup function will be called as part of `~ApiHandle()` destructor.

```
-                SecureTunnelingFeature::~SecureTunnelingFeature() { aws_http_library_clean_up(); }
+                SecureTunnelingFeature::~SecureTunnelingFeature() = default;
```

- Prior to this change, I observed that there were 2 references to the SharedCrtResourceManager instance at the time of `shutdown()` (see output of `use_count()` below). This is unexpected and there should only be 1 reference after all features have stopped.  The extra reference prevented the destructor `~SharedCrtResourceManager()` from being invoked and as a result, the `~ApiHandle()` destructor was never invoked.

```
(gdb) p resourceManager
$1 = std::shared_ptr<Aws::Iot::DeviceClient::SharedCrtResourceManager> (use count 2, weak count 0) = {get() = 0x557c187a3330}
```

After this change, I observed exactly 1 reference to SharedCrtResourceManager after all features have stopped. 

```
Thread 1 "aws-iot-device-" hit Breakpoint 1, shutdown () at /src/source/main.cpp:150
150         resourceManager.reset();
(gdb) p resourceManager
$1 = std::shared_ptr<Aws::Iot::DeviceClient::SharedCrtResourceManager> (use count 1, weak count 0) = {get() = 0x55f6adb77bc0}
```

The single reference can now be released just prior to calling `exit()` and the `~ApiHandle()` destructor will be invoked. Please add the line `resourceManager.reset();` to the `shutdown()` function as shown below.

```
void shutdown()
{
    ...
    LoggerFactory::getLoggerInstance().get()->shutdown();
    resourceManager.reset();
    exit(0);
}
```

## Resolution
1. Modify `~SecureTunnelingFeature()` destructor to remove call to `aws_http_library_clean_up()`.
2. Modify `shutdown()` function to explicit release `SharedCrtResourceManager` just prior to `exit(0)` by calling `resourceManager.reset();`.

**Review by shangabl — APPROVED** [review]

**Review by marcoemorais-aws — APPROVED** [review]
👍
